dirty_cat/datetime_encoder.py

```diff
@@ -0,0 +1,173 @@
+from pandas.tseries.holiday import USFederalHolidayCalendar as calendar
+from typing import Union, Optional, List
+
+
+
+class DatetimeEncoder(TransformerMixin, BaseEstimator):
+    """
+    This encoder transforms each datetime column into several numeric columns corresponding to temporal features,
+    e.g year, month, day...
+    If one of the feature is constant, this feature is dropped.
+    If the dates are timezone aware, all the features extracted will correspond to the provided timezone.
+
+    Parameters
+    ----------
+    extract_until : {"year", "month", "day", "hour", "minute", "second",
+    "millisecond", "microsecond", "nanosecond"}, default="hour"
+        Extract up to this granularity, and gather the rest into the "other" feature.
+        For instance, if you specify "day", only "year", "month", "day" and "other" features will be created.
+        The "other" feature will be a numerical value expressed in the "extract_until" unit.
+    add_day_of_the_week: bool, default=False
+        Add day of the week feature (if day is extracted). This is a numerical feature from 0 to 6.
+    add_holidays : bool, default=False
+        Whether to add a numerical variable encoding if the day of the date is a holiday (1 for holiday,
+        0 for non-holiday).
+        Uses pandas calendar, which for now only contains US holidays.
+
+    Attributes
+    ----------
+    n_features_out: int
+        Number of features of the transformed data.
+    features_per_column: Dict[int, List[str]]
+        A dictionary mapping the index of the original columns
+        to the list of features extracted for each column.
+    """
+    def __init__(self,
+                 extract_until="hour",
+                 add_day_of_the_week=False,
+                 add_holidays=False):
+        to_extract = ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]
+        self.extract_until = extract_until
+        self._validate_keywords()
+        self.to_extract_full = to_extract[:to_extract.index(extract_until) + 1]
+        self.to_extract_full.append("other")
+        self.add_day_of_the_week = add_day_of_the_week
+        self.add_holidays = add_holidays
+        # Some functions need aliases
+        self.word_to_alias = {"year": "Y", "month": "M", "day": "D", "hour": "H", "minute": "min", "second": "S",
+                              "millisecond": "ms", "microsecond": "us", "nanosecond": "N"}
+
+    def _validate_keywords(self):
+        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
+            msg = (
+                """extract_until should be one of ["year", "month", "day", "hour", "minute", "second", "millisecond", 
+                "microsecond", "nanosecond"], got {0}.""".format(
+                    self.extract_until
+                )
+            )
+            raise ValueError(msg)
```

Comment:
I'd advise using modern python f-strings, which are usually nicer to read :)

```suggestion
        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
            raise ValueError(
                f'"extract_until" should be one of ["year", "month", "day", "hour", "minute", '
                f'"second", "millisecond", "microsecond", "nanosecond"], got {self.extract_until}.'
            )
```